### DISCUSSION TITLE
Where does PythonOperator python_callable get its context kwargs from?

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#16152] [Q&A] [answered]
Hi there!

I'm trying to understand how a PythonOperator's python_callable receives the airflow-related information - namely `dag_run`, `ti`, `dag`, `conf` and friends.

After reading the code base for a good hour, I did get some progress in, but I'm still stumped.
My understanding is that the variables above are created/gathered in [`airflow.models.taskinstance.get_template_context`](https://github.com/apache/airflow/blob/afa1723d2493fa23676f1b3d34c8b71f41d7503f/airflow/models/taskinstance.py#L1558), but the implementation of PythonOperator does not have anywhere that calls the `get_template_context` function, nor does it seem to make any call to super that would update the python_callable args. Observations are made as comments below:

*`PythonOperator.__init__`*:

```py
def __init__(
        self,
        *,
        python_callable: Callable,
        op_args: Optional[List] = None,
        op_kwargs: Optional[Dict] = None,
        templates_dict: Optional[Dict] = None,
        templates_exts: Optional[List[str]] = None,
        **kwargs,
    ) -> None:
        if kwargs.get("provide_context"):
            warnings.warn(
                "provide_context is deprecated as of 2.0 and is no longer required",
                DeprecationWarning,
                stacklevel=2,
            )
            kwargs.pop('provide_context', None)
        super().__init__(**kwargs)
        if not callable(python_callable):
            raise AirflowException('`python_callable` param must be callable')
        self.python_callable = python_callable # straight from the __init__ args
        self.op_args = op_args or []
        self.op_kwargs = op_kwargs or {} # no modification done here, or demonstration that op_kwargs is obtained from a parent class
        self.templates_dict = templates_dict
        if templates_exts:
            self.template_ext = templates_exts       
```

* Did not find a `__post_init__` in Parent classes
* Did not find any behaviour that seemed to modify the `__init__` args in  `BaseOperator.__new__`
* `BaseOperator.__init__` does not make ref to `op_kwargs`, and even if it did, it's overriden by `self.op_kwargs = op_kwargs or {}`

*`PythonOperator.execute`*:

```py
def execute(self, context: Dict):
        # no super call, + BaseOperator does not implement execute
        context.update(self.op_kwargs) # I couldn't find refs to op_kwargs in BaseOperator, so I assume this is the same as seen in __init__
        context['templates_dict'] = self.templates_dict # OK, one extra property, but it's not like it's spread into the dict

        self.op_kwargs = determine_kwargs(self.python_callable, self.op_args, context) # from my reading, determin_kwargs seems to be a filter of the kwargs, so no addition or spreading of template_dict

        return_value = self.execute_callable()
        self.log.info("Done. Returned value was: %s", return_value)
        return return_value
```

*`PythonOperator.execute_callable`*:

```py
def execute_callable(self):
        """
        Calls the python callable with the given arguments.

        :return: the return value of the call.
        :rtype: any
        """
        # again, no clear sign of call to super or modification for op_args or op_kwargs, so I expect both to be the same as in execute
        return self.python_callable(*self.op_args, **self.op_kwargs)
```

I was unable to find where the template_context is integrated in the arguments of the python_callable. Where is this arcane magic performed? How can I know what's the full signature, and what's minimally required?

Thanks a lot for your answers, you wizards! 🧙 🙏

**Comment by turbaszek** [comment, marked as answer]
You are correct that the answer is in `context`.

`context` has all Airflow information (task_id, dag_run etc) and it is the `context` that is being first updated with arguments passed by users and later on passed to the callable.

https://github.com/apache/airflow/blob/8d16638285687fd0ef41d40340ab1c5bcffd507a/airflow/operators/python.py#L145
Here we update airflow-related dictionary with airflow-unrelated info (arguments to task callable) and that is where the magic happens 🧙  Doing it this way gives users possibility to override the execution context. Then we merged this context with `self.op_args` here:
https://github.com/apache/airflow/blob/8d16638285687fd0ef41d40340ab1c5bcffd507a/airflow/operators/python.py#L148
We do this because arguments to callable can be passed either via `args` or `kwargs`. Finally we use this composite `op_kwargs` (both airflow and users input!) as input to the python callable:
https://github.com/apache/airflow/blob/8d16638285687fd0ef41d40340ab1c5bcffd507a/airflow/operators/python.py#L161

In this way all Airflow related information is being passed to the callable and is passed only if the callable accepts it (explicit keyword or via `**kwargs`) because of `determine_kwargs`.

## More context of the `context`
The story of a task ends here where we pass the context during execution:
https://github.com/apache/airflow/blob/e7d3fa2dbd8296a95dbd7bb60adbd148d7336285/airflow/models/taskinstance.py#L1329

From that place you can try to track the origin of `context` and finally you will get to definition of Airflow context:
https://github.com/apache/airflow/blob/e7d3fa2dbd8296a95dbd7bb60adbd148d7336285/airflow/models/taskinstance.py#L1684-L1724

**Reply by p-hebert** [reply]
Lovely, thanks @turbaszek! That's a great explanation and explains a lot of the more arcane parts :)

The only thing I still don't understand is how the `context` passed to the `PythonOperator.execute` function contains the airflow information when execute can be (and sometimes is!) called directly by client code, i.e.:

```
PythonOperator(task_id="task_id", python_callable=my_python_callable).execute(context={})
```

How does this work?

**Reply by potiuk** [reply]
It works exactly like this. Context is empty dict in this case and you cannot expect anything more. There should never be reason to call PythonOperator's.execute() directly (you can in tests when you mock execution). It can be called if your operator derives from PythonOperator and overrides execute() so it can calls super().execute() but it should pass the context that you received from your execute in this context.

**Reply by p-hebert** [reply]
Great, thanks for the precision @potiuk! That solves it for me. We do call some `.execute()` manually on our side when we have more dynamic DAGs, but I'll take note that this is bad practice :)

That solves this issue! 🎉
